### ISSUE TITLE
masked_select hangs indefinitely with large tensor inputs

### ISSUE THREAD

**Issue by L1aoXingyu** [geohot/tinygrad#10340]
When using the tiny backend for HLB-CIFAR10 torch training, I encountered a hang issue upon launching the training script. By isolating each component and testing individually, I discovered the issue originates from the masked_select function.

Further examination revealed that torch.masked_select internally calls tinygrad's masked_select. Testing this independently confirmed the problem occurs when processing large tensor sizes, causing the operation to hang indefinitely.

Below is a simplified reproduction test case:

```python
from tinygrad import Tensor, dtypes

def make_random_square_masks(inputs: Tensor, mask_size: int):
    if mask_size == 0:
        return None
    is_even = int(mask_size % 2 == 0)
    in_shape = inputs.shape

    mask_center_y = Tensor.randint(in_shape[0], low=mask_size//2-is_even, high=in_shape[-2]-mask_size//2-is_even + 1, device=inputs.device, dtype=dtypes.long)
    mask_center_x = Tensor.randint(in_shape[0], low=mask_size//2-is_even, high=in_shape[-1]-mask_size//2-is_even + 1, device=inputs.device, dtype=dtypes.long)

    # measure distance, using the center as a reference point
    to_mask_y_dists = Tensor.arange(in_shape[-2], device=inputs.device, dtype=dtypes.long).reshape(1, 1, in_shape[-2], 1) - mask_center_y.reshape(-1, 1, 1, 1)
    to_mask_x_dists = Tensor.arange(in_shape[-1], device=inputs.device, dtype=dtypes.long).reshape(1, 1, 1, in_shape[-1]) - mask_center_x.reshape(-1, 1, 1, 1)

    to_mask_y = (to_mask_y_dists >= (-(mask_size // 2) + is_even)) * (to_mask_y_dists <= mask_size // 2)
    to_mask_x = (to_mask_x_dists >= (-(mask_size // 2) + is_even)) * (to_mask_x_dists <= mask_size // 2)
    final_mask = to_mask_y * to_mask_x ## Turn (y by 1) and (x by 1) boolean masks into (y by x) masks through multiplication. Their intersection is square, hurray! :D

    return final_mask


B, C, H_img, W_img = 10000, 3, 36, 36 # Batch, Channels, Padded Height, Padded Width
crop_size = 32
input_tensor = Tensor.randn(B, C, H_img, W_img).half()

mask = make_random_square_masks(input_tensor, crop_size)

cropped_inputs = Tensor.masked_select(input_tensor, mask)

print(cropped_inputs.numpy())
```

In the above test, setting B=10000 results in an indefinite hang with no output. However, when using a smaller batch size (B=1000), the code executes successfully and produces output. This clearly indicates an issue within masked_select's handling of large input sizes.

**Comment by L1aoXingyu** [comment]
simpler reproduction case

```python
B, C, H, W = 50000, 3, 36, 36
x     = Tensor.randn(B, C, H, W).half()
mask  = Tensor.rand(B, C, H, W) > 0.5
y     = Tensor.masked_select(x, mask)
print(y.numpy())
```

**Comment by IntendedConsequence** [comment]
I've hit the exact same problem in #10728. This happens with any sufficiently large tensor indexing, doesn't matter if you use indexing `x[indices]`, `Tensor.gather`, `Tensor.masked_select` or whatever, because all of them go through a big generic `Tensor._getitem` that emits what looks like a fairly complicated masks computation logic even for very simple one-dimensional indexing, which doesn't seem to get optimized away with default settings.
FUSE_ARANGE=1 seems to fix this lag however. Running the code below (I tested on CUDA) should reproduce the issue.

Without FUSE_ARANGE:
```
loading 159.96 ms
indexing 4271.95 ms
```

With FUSE_ARANGE=1:
```
loading 158.01 ms
indexing  54.97 ms
```

Switching from `Tensor.arange` to `Tensor(list(range(X.shape[0])))` doesn't make a difference, showing that FUSE_ARANGE=1 doesn't only affect `Tensor.arange`.
But if we switch to `Tensor.randperm` as a source for random indices, loading time increases to 1800ms regardless of FUSE_ARANGE, while indexing only increases by +25ms or so, with FUSE_ARANGE=1.

```python
from tinygrad import Tensor, Device
from tinygrad.helpers import Timing, getenv

N = getenv("N", 50000)
M = getenv("M", 3*32*32)

with Timing(f"loading "):
    X = Tensor.rand(N, M).realize()
    order = Tensor.arange(X.shape[0]).realize()
    # order = Tensor(list(range(X.shape[0]))).realize()
    Device[Device.DEFAULT].synchronize()

with Timing(f"indexing "):
    Y = X[order].realize()
    Device[Device.DEFAULT].synchronize()
print(Y[0].numpy())
```
